Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/Mandate/Notification.hs

```diff
@@ -0,0 +1,105 @@
+
+import qualified Data.Map.Strict as Map
+import Domain.Types.DriverFee as DF
+import Domain.Types.DriverInformation as DI
+import Domain.Types.DriverPlan as DPlan
+import Domain.Types.Mandate (Mandate)
+import qualified Domain.Types.Notification as NTF
+import Domain.Types.Person as P
+import Domain.Types.Plan as Plan
+import qualified Kernel.External.Payment.Interface.Types as PaymentInterface
+import Kernel.Prelude
+import qualified Kernel.Storage.Esqueleto as Esq
+import Kernel.Types.Error
+import Kernel.Types.Id (Id, cast)
+import Kernel.Utils.Common
+import qualified Lib.Payment.Domain.Action as APayments
+import Lib.Scheduler
+import SharedLogic.Allocator
+import qualified Storage.Queries.DriverFee as QDF
+import qualified Storage.Queries.DriverInformation as QDI
+import qualified Storage.Queries.DriverPlan as QDP
+import qualified Storage.Queries.Notification as QNTF
+import qualified Tools.Payment as TPayment
+
+sendPDNNotificationToDriver ::
+  ( CacheFlow m r,
+    EsqDBFlow m r,
+    Esq.EsqDBReplicaFlow m r,
+    EncFlow m r
+  ) =>
+  Job 'SendPDNNotificationToDriver ->
+  m ExecutionResult
+sendPDNNotificationToDriver Job {id, jobInfo} = withLogTag ("JobId-" <> id.getId) do
+  let jobData = jobInfo.jobData
+      startTime = jobData.startTime
+      endTime = jobData.endTime
+  activeSubscribedDrivers <- QDI.findAllSubscribedByAutoPayStatusAndMerchantId jobData.merchantId (Just DI.ACTIVE) True
+  mandateIdAndDriverIdsToNotify <- mandateIdAndDriverId <$> QDP.findAllByDriverIdsAndPaymentMode (DI.driverId <$> activeSubscribedDrivers) Plan.AUTOPAY
+  driverFees <- QDF.findFeesInRangeAndDriverIdsWithStatus startTime endTime DF.PAYMENT_PENDING (fst <$> mandateIdAndDriverIdsToNotify)
+  let driverInfoForPDNotification = mapDriverInfoForPDNNotification (Map.fromList mandateIdAndDriverIdsToNotify) driverFees
+  for_ driverInfoForPDNotification $ \driverToNotify -> do
+    notificationId <- generateGUID
+    notificationShortId <- generateShortId
+    now <- getCurrentTime
+    req <- mkNotificationRequest driverToNotify notificationShortId.getShortId
+    exec <- try @_ @SomeException (APayments.createNotificationService req (TPayment.mandateNotification jobData.merchantId))
+    case exec of
+      Left _ -> throwError (InternalError $ "Notificaation failed for driverFeeId" <> driverToNotify.driverFeeId.getId)
+      Right res -> QNTF.create $ buildNotificationEntity res notificationId driverToNotify.driverFeeId driverToNotify.mandateId now
+  return Complete
+  where
+    mandateIdAndDriverId =
+      mapMaybe
+        ( \dplan ->
+            case DPlan.mandateId dplan of
+              Just mandateId_ -> Just (dplan.driverId, mandateId_)
+              Nothing -> Nothing
+        )
+    mapDriverInfoForPDNNotification mapMandateByDriverId =
+      mapMaybe
+        ( \driverFee -> mkInfoForPDNNotification driverFee <$> (mapMandateByDriverId Map.!? (cast @P.Driver @P.Person driverFee.driverId))
```

Comment:
Why map when you can traverse the list itself? This is not very safe and is used in execution too

Reply:
Bro just wanted to omit drivers that are not eligible for notification (that's why used `mapMaybe`) similarly in case of execution.

Reply:
But sure will change it.